### PR TITLE
Run Unit Test in Travis builds

### PR DESCRIPTION
**Pull request by stephengroat**
`osx` in Travis has `brew` available without `sudo`, easy to install `cmocka`

For `linux` and appveyor, I created a small script to download and compile `cmocka`.

Unit test `test_tb_x86_64_32_imul_Gv_Ev_Ib` now failing on all platforms

TODO: see if `clang` works in `cygwin` for builds, currently following up with cmocka

What about changing the `tests/unit` to pass and adding it to the build process? Maybe `bindings` too?

### CODE REVIEW

**Comment by aquynh** [comment]
great, please keep doing these experiments, and ping us when you are done! i know it is very tricky & tedious to make this work, especially with appveyor

**Comment by stephengroat** [comment]
that's what beer is for!

**Comment by stephengroat** [comment]
I've started talking with the appveyor folks about the problems with cmake in cygwin (needed for cmocka).

While that's going on, can anyone comment on why the `linux` and `osx` unit tests are failing?

**Comment by stephengroat** [comment]
Last few builds tested python binding, those are also failing the sample `diff`s

**Comment by danghvu** [comment]
I would skip the binding tests from CI for now, there is too often problems of output format mismatch. The tests there are mostly for the binding developer to use as reference.

**Comment by stephengroat** [comment]
Bindings are second now, we can always quickly remove later. current build failures are from `unit/test`

**Comment by stephengroat** [comment]
appveyor and windows now builds and runs tests (failing tests, like all other platforms) for cygwin and cygwin64, both using gcc

**Comment by stephengroat** [comment]
@aquynh i now have the unit test `test_tb_x86_64_32_imul_Gv_Ev_Ib` failing on all platforms in the same place (expect cygwin `clang`, still working on getting that to compile). can you take a look? i'm not 100% sure what's going on there.

**Comment by aquynh** [comment]
Yes, that is an unsolved case, so you can ignore it.

**Comment by stephengroat** [comment]
@aquynh could you push to the branch to disable the test? i'm not 100% sure how to do it!

File: samples/Makefile [hunk 1]
```diff
@@ -20,7 +21,7 @@ SAMPLEDIR = .
 OBJDIR = .
 LIBDIR = ..
 
-CFLAGS += -Wall -I$(INCDIR)
+CFLAGS += -Wall -Werror -Wno-unused-function -I$(INCDIR)
```
Comment (aquynh) [inline samples/Makefile]: if there is unused function, is it better to just clean it, rather than disable the warning?

**Comment by aquynh** [comment]
i just have some minor comments. very nice work!

**Comment by aquynh** [comment]
to disable that test_tb_x86_64_32_imul_Gv_Ev_Ib, you can comment out the line with `./test_tb_x86` in `unit/Makefile`

File: Makefile [hunk 2]
```diff
@@ -100,65 +101,63 @@ VERSION_EXT =
 
 BIN_EXT =
 
-IS_APPLE := $(shell $(CC) -dM -E - < /dev/null | grep -cm 1 -e __apple_build_version__ -e __APPLE_CC__)
-ifeq ($(IS_APPLE),1)
+# Apple?
+ifeq ($(UNAME_S),Darwin)
```
Comment (stephengroat) [inline Makefile]: Compiling on iOS or cross compiling for iOS on macOS?

File: samples/Makefile (hunk 1 above)
Comment (stephengroat) [inline samples/Makefile]: agreed. cleaned

**Comment by stephengroat** [comment]
new test failing, `test_x86_soft_paging`

**Comment by aquynh** [comment]
yes i remember test_x86_soft_paging is another unsolved case.

**Comment by stephengroat** [comment]
ok, i'll comment that out too.

i'm currently trying to get `macos_universal` building in a reasonable amount of time by caching the `brew install glib --universal` output (was thinking about making the `universal` the default?).

**Comment by aquynh** [comment]
i think in most system glib is not compiled in universal mode, so we cannot compile with universal as default option.

**Comment by stephengroat** [comment]
ok. reading the comments of `test_hang`, I'm guessing that's one to skip too?

**Comment by stephengroat** [comment]
trying to integrate #574 and #647 into the `osx` builds

for #647, would `brew unlink` allow me to install `glib` and `glib --universal` side by side with brew?

File: Makefile (hunk 2 above)
Comment (aquynh) [inline Makefile]: i mean cross compile from Mac, but then it should work.

**Comment by aquynh** [comment]
is this done yet?

**Comment by stephengroat** [comment]
depends on what you're looking for. what tests are you looking to run?

`test_hang` seems to be failing on `clang` for `osx` and `linux` and `gcc` for `windows` and `test_x86_shl_enter_leave` fails for all

should both of those be commented out?

**Comment by aquynh** [comment]
yes please comment them out for now. i will look at them later.

**Comment by stephengroat** [comment]
squashed into a single commit that builds, will do separate PRs for some of the other stuff (mingw, bindings, etc.)

**Comment by stephengroat** [comment]
Any ideas on the failing cross compiles in the latest build?

`pthread` error? seen it before?

**Comment by aquynh** [comment]
are you done with this yet?

**Comment by stephengroat** [comment]
yup. cleaned up commits and ready to push. comments still in place

File: tests/unit/test_tb_x86.c [hunk 3]
```diff
@@ -222,7 +222,7 @@ static void test_tb_x86_64_32_imul_Gv_Ev_Ib(void **state)
 {
     uc_engine *uc = *state;
     uc_hook trace1, trace2;
-    void *mem;
+    //void *mem;
```
Comment (danghvu) [inline tests/unit/test_tb_x86.c]: What is this for ? If we exclude this test, let's leave this out from this commit.

Comment (stephengroat) [inline tests/unit/test_tb_x86.c]: Doesn't compile with it in

Comment (danghvu) [inline tests/unit/test_tb_x86.c]: I don't understand, which one doesn't compile with it and what is the error ?

File: tests/unit/test_tb_x86.c [hunk 4]
```diff
@@ -245,8 +245,9 @@ static void test_tb_x86_64_32_imul_Gv_Ev_Ib(void **state)
     int64_t edi = 0x488ac239;
 #endif
 
-    mem = calloc(1, CODE_SPACE);
-    assert_int_not_equal(0, mem);
+    //mem = calloc(1, CODE_SPACE);
+    // TODO examine
+    //assert_int_not_equal(0, mem);
```
Comment (stephengroat) [inline tests/unit/test_tb_x86.c]: @danghvu referring to your comment on line 255, this assert is the issue. since `mem` is a `void*` and not an `int`, this doesn't compile. with `-Wall -Werror`, you then need to remove an unused variable

**Comment by aquynh** [comment]
this looks great! for the record, can you summarize all the changes of this PR?

**Comment by stephengroat** [comment]
Summary:
- in appveyor, install clang and cmake in cygwin, enable package upgrades, and build cmocka and enable testing for gcc only
- in `gitignore`, ignore generated cmocka folder
- in travis, use brew in osx to install cmocka, and enable testing for gcc and clang on os x and linux
- in `Makefile`, change to use `uname -s` to determine os type
- make `install-cmocka-linux.sh`, a simple shell script to download and install cmocka on linux
- in `bindings/Makefile`, enable `make -c` to call subdirectory makefiles instead of `cd [dir] && make` and include environment variables for runtime access to generated libraries
- in `samples/Makefile`, change to use `uname -s` to determine os type, remove `clean_bins` from `all` command, and include `Werror` for compile strictness
- in `tests/unit/Makefile`, add `cflags` for compile time access to cmocka headers and library, include execute vars for runtime access to cmocka and unicorn libs
- in `tests/unit/test_tb_x86.c`, comment out assert that would not compile 

I can move it to the commit message if necessary

**Comment by aquynh** [comment]
merged, thanks for your fantastic work: https://github.com/unicorn-engine/unicorn/commit/867cd92c406cafe0117272d66c8733e86946fbb6!

**Comment by stephengroat** [comment]
can you try triggering a re-run on travis?

**Comment by aquynh** [comment]
Ok Travis passed now, that was false positive, thanks.

**Comment by stephengroat** [comment]
looking at it

**Comment by stephengroat** [comment]
appveyor is unhappy too

**Comment by aquynh** [comment]
Appveyor had no complain in the latest build.
